lib/web_ui/lib/src/engine/platform_dispatcher/app_lifecycle_state.dart

```diff
@@ -0,0 +1,86 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:ui/ui.dart' as ui;
+
+import '../../engine.dart';
+
+/// Signature of functions added as a listener to [ui.AppLifecycleState] changes
+typedef AppLifecycleStateListener = void Function(ui.AppLifecycleState state);
+
+/// Determines the [ui.AppLifecycleState].
+abstract class AppLifecycleState {
+  static final AppLifecycleState instance = _BrowserAppLifecycleState();
+
+  ui.AppLifecycleState get appLifecycleState => _appLifecycleState;
+  ui.AppLifecycleState _appLifecycleState = ui.AppLifecycleState.resumed;
+
+  final List<AppLifecycleStateListener> _listeners =
+      <AppLifecycleStateListener>[];
+
+  void addListener(AppLifecycleStateListener listener) {
+    if (_listeners.isEmpty) {
+      activate();
+    }
+    _listeners.add(listener);
+    listener(_appLifecycleState);
+  }
+
+  void removeListener(AppLifecycleStateListener listener) {
+    _listeners.remove(listener);
+    if (_listeners.isEmpty) {
+      deactivate();
+    }
+  }
+
+  @protected
+  void activate();
+
+  @protected
+  void deactivate();
+
+  @visibleForTesting
+  void onAppLifecycleStateChange(ui.AppLifecycleState newState) {
+    if (newState != _appLifecycleState) {
+      _appLifecycleState = newState;
+      for (final AppLifecycleStateListener listener in _listeners) {
+        listener(newState);
+      }
+    }
+  }
+}
+
+/// Determines the [ui.AppLifecycleState] by listening to 'beforeunload',
+/// 'visibilitychange' events on the [DomWindow] / [DomHTMLDocument].
+class _BrowserAppLifecycleState extends AppLifecycleState {
+  @override
+  void activate() {
+    domWindow.addEventListener('beforeunload', _beforeUnloadListener);
+    domDocument.addEventListener('visibilitychange', _visibilityChangeListener);
```

Comment:
Why isn't there also a listener here for `windows.onFocusChanged` that sets the lifecycle state to `inactive`? This is a useful state for applications, since they can unfocus widgets, close menus, or otherwise indicate that they no longer have input focus.